File: backend/app/routers/categorias.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user, require_central
from app.models.categoria import Categoria
from app.schemas.categoria import CategoriaCreate, CategoriaOut, CategoriaUpdate

router = APIRouter()


@router.get("/", response_model=List[CategoriaOut])
def listar_categorias(
    db: Session = Depends(get_db),
    _=Depends(get_current_user),
):
    return db.query(Categoria).order_by(Categoria.nome).all()


@router.post("/", response_model=CategoriaOut, status_code=201)
def criar_categoria(
    dados: CategoriaCreate,
    db: Session = Depends(get_db),
    _=Depends(require_central),
):
    categoria = Categoria(**dados.model_dump())
    db.add(categoria)
    db.commit()
    db.refresh(categoria)
    return categoria


@router.put("/{categoria_id}", response_model=CategoriaOut)
def atualizar_categoria(
    categoria_id: int,
    dados: CategoriaUpdate,
    db: Session = Depends(get_db),
    _=Depends(require_central),
):
    categoria = db.query(Categoria).filter(Categoria.id == categoria_id).first()
    if not categoria:
        raise HTTPException(status_code=404, detail="Categoria não encontrada")
    for campo, valor in dados.model_dump(exclude_unset=True).items():
        setattr(categoria, campo, valor)
    db.commit()
    db.refresh(categoria)
    return categoria


@router.delete("/{categoria_id}", status_code=204)
def deletar_categoria(
    categoria_id: int,
    db: Session = Depends(get_db),
    _=Depends(require_central),
):
    categoria = db.query(Categoria).filter(Categoria.id == categoria_id).first()
    if not categoria:
        raise HTTPException(status_code=404, detail="Categoria não encontrada")
    db.delete(categoria)
    db.commit()

File: frontend/src/api/categorias.ts
import api from "./client"

export interface Categoria {
  id: number
  nome: string
  descricao: string | null
}

export const getCategorias = () => api.get<Categoria[]>("/categorias/").then(r => r.data)
export const criarCategoria = (dados: { nome: string; descricao?: string }) => api.post<Categoria>("/categorias/", dados).then(r => r.data)
export const atualizarCategoria = (id: number, dados: { nome?: string; descricao?: string }) => api.put<Categoria>(`/categorias/${id}`, dados).then(r => r.data)
export const deletarCategoria = (id: number) => api.delete(`/categorias/${id}`)
